packages/devtools_app/lib/src/screens/performance/rebuild_counts.dart

```diff
@@ -20,112 +21,297 @@ import '../../primitives/utils.dart';
-          (json['newLocations'] as Map).cast<String, dynamic>();
+    }
 
-      for (final file in fileLocationsMap.keys) {
-        final List<int> entries = (fileLocationsMap[file] as List).cast<int>();
+    _locationsResolved.value = _countUnkownLocations == 0;
+  }
+}
 
-        final shortName = path.posix.split(file).last;
+class RebuildCountModel {
+  RebuildCountModel();
 
-        final len = entries.length ~/ 3;
-        for (var i = 0; i < (len * 3); i += 3) {
-          final id = entries[i];
-          final line = entries[i + 1];
-          final location = RebuildLocation(
-            id: id,
-            path: file,
-            line: line,
-            column: entries[i + 2],
-            name: '$shortName:$line',
-          );
-          _locationMap[id] = location;
-          _locations.add(location);
-        }
+  RebuildCountModel.fromJson(Map<String, dynamic> json) {
+    if (json.isEmpty) return;
+    locationMap.processLocationMap(json['locations']);
+    final frames =
+        (json['frames'] as List<dynamic>).cast<Map<String, dynamic>>();
+    frames.forEach(processRebuildsForFrame);
+  }
+
+  // Maximum number of historic frames to keep rebuild counts to ensure memory
+  // usage from rebuild counts is not excessive.
+  static const int rebuildFrameCacheSize = 10000;
+
+  /// Source of truth for all resolution fo location ids to [Location] objects.
+  final LocationMap locationMap = LocationMap();
+
+  /// Map from frame id to list of rebuilds for the frame.
+  final _rebuildsForFrame = <int, List<RebuildLocation>>{};
+
+  final _stats = _RebuildCountStats();
+  ValueListenable<List<RebuildLocation>> get locationStats => _stats.locations;
+
+  List<RebuildLocation>? get rebuildsForLastFrame =>
+      _rebuildsForFrame.values.lastOrNull;
+
+  List<RebuildLocation>? rebuildsForFrame(int frameNumber) {
+    return _rebuildsForFrame[frameNumber];
+  }
+
+  Map<String, dynamic>? toJson() {
+    if (_rebuildsForFrame.isEmpty) {
+      // No need to encode data unless there were actually rebuilds reported.
+      return null;
+    }
+    final frames = <dynamic>[];
+
+    _rebuildsForFrame.forEach((id, rebuilds) {
+      final events = <int>[];
+      for (RebuildLocation rebuild in rebuilds) {
+        events
+          ..add(rebuild.location.id)
+          ..add(rebuild.buildCount);
       }
+      frames.add({'frameNumber': id, 'events': events});
+    });
+    return <String, dynamic>{
+      'locations': locationMap.toJson(),
+      'frames': frames,
+    };
+  }
+
+  void processRebuildEvent(Map<String, dynamic> json) {
+    // parse locations
+    if (json.containsKey('locations')) {
+      locationMap.processLocationMap(json['locations'] as Map<String, dynamic>);
     }
 
+    processRebuildsForFrame(json);
+
+    // We've updated the build counts and possibly the locations.
+    _stats.dataChanged();
+  }
+
+  void clearFromRestart() {
+    _rebuildsForFrame.clear();
+    locationMap.clear();
+    _stats.clear();
```

Comment:
nit: rewrite as follows to reduce duplication
```
locationMap.clear();
clearAllCounts();
```

Reply:
done although reversed order to be idiomatic so future changes to clearAllCounts don't run into trouble because the locationMap was cleared first.